File: src/mnemonic/cache.py
"""Cache module for Mnemonic."""

from __future__ import annotations

import os
import platform
import shutil
from dataclasses import dataclass
from datetime import datetime, timedelta
from pathlib import Path
from typing import Protocol


@dataclass(frozen=True)
class CacheInfo:
    """キャッシュ情報"""

    directory: Path
    size_bytes: int
    template_version: str | None
    template_expires_in_days: int | None


class CacheManager(Protocol):
    """キャッシュ管理インターフェース"""

    def get_cache_dir(self) -> Path:
        """OSごとのキャッシュディレクトリを取得する"""
        ...

    def get_template_cache_path(self, version: str) -> Path:
        """テンプレートキャッシュパスを取得する"""
        ...

    def is_cache_valid(self, path: Path, max_age_days: int) -> bool:
        """キャッシュの有効性をチェックする"""
        ...

    def clear_cache(self, template_only: bool = False) -> None:
        """キャッシュをクリアする"""
        ...

    def get_cache_info(self) -> CacheInfo:
        """キャッシュ情報を取得する"""
        ...


def get_cache_dir() -> Path:
    """OSごとのキャッシュディレクトリを取得する"""
    system = platform.system()

    if system == "Linux":
        xdg_cache = os.environ.get("XDG_CACHE_HOME")
        base = Path(xdg_cache) if xdg_cache else Path.home() / ".cache"
    elif system == "Darwin":
        base = Path.home() / "Library" / "Caches"
    elif system == "Windows":
        local_app_data = os.environ.get("LOCALAPPDATA")
        if local_app_data:
            base = Path(local_app_data) / "mnemonic"
        else:
            base = Path.home() / "AppData" / "Local" / "mnemonic"
        return base / "cache"
    else:
        base = Path.home() / ".cache"

    return base / "mnemonic"


def get_template_cache_path(version: str) -> Path:
    """テンプレートキャッシュパスを取得する"""
    return get_cache_dir() / "templates" / version


def is_cache_valid(path: Path, max_age_days: int = 7) -> bool:
    """キャッシュの有効性をチェックする"""
    if not path.exists():
        return False

    mtime = datetime.fromtimestamp(path.stat().st_mtime)
    age = datetime.now() - mtime
    return age < timedelta(days=max_age_days)


def clear_cache(template_only: bool = False) -> None:
    """キャッシュをクリアする"""
    cache_dir = get_cache_dir()

    if template_only:
        template_dir = cache_dir / "templates"
        if template_dir.exists():
            shutil.rmtree(template_dir)
    else:
        if cache_dir.exists():
            shutil.rmtree(cache_dir)


def get_cache_info() -> CacheInfo:
    """キャッシュ情報を取得する"""
    cache_dir = get_cache_dir()

    if not cache_dir.exists():
        return CacheInfo(
            directory=cache_dir,
            size_bytes=0,
            template_version=None,
            template_expires_in_days=None,
        )

    total_size = sum(f.stat().st_size for f in cache_dir.rglob("*") if f.is_file())

    template_dir = cache_dir / "templates"
    template_version = None
    template_expires = None

    if template_dir.exists():
        versions = list(template_dir.iterdir())
        if versions:
            latest = max(versions, key=lambda p: p.stat().st_mtime)
            template_version = latest.name
            mtime = datetime.fromtimestamp(latest.stat().st_mtime)
            expires = 7 - (datetime.now() - mtime).days
            template_expires = max(0, expires)

    return CacheInfo(
        directory=cache_dir,
        size_bytes=total_size,
        template_version=template_version,
        template_expires_in_days=template_expires,
    )

File: internal/cache/cache.go
// Package cache はキャッシュディレクトリ（テンプレート、SDL2ソース、デバッグ用署名鍵）の
// 解決と管理を提供する。
package cache

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"math"
	"os"
	"path/filepath"
	"runtime"
	"time"
)

// TemplateMetadataFilename はテンプレートキャッシュのメタデータファイル名。
const TemplateMetadataFilename = "metadata.json"

// TemplateMetadataTimeLayout はメタデータ内の日時フォーマット（UTC）。
const TemplateMetadataTimeLayout = "2006-01-02T15:04:05Z"

// TemplateMetadata はテンプレートキャッシュ1バージョンのメタデータ。
type TemplateMetadata struct {
	Version      string `json:"version"`
	DownloadedAt string `json:"downloaded_at"`
	ExpiresAt    string `json:"expires_at"`
}

// KeystoreDirName はデバッグ用署名鍵を置くキャッシュ配下のサブディレクトリ名。
const KeystoreDirName = "keystore"

// Info はキャッシュディレクトリの情報を表す。
//
// TemplateVersion / TemplateExpiresInDays はテンプレートキャッシュが存在しない場合の
// 未設定を表現するため、ポインタとする。
type Info struct {
	Directory             string
	SizeBytes             int64
	TemplateVersion       *string
	TemplateExpiresInDays *int
}

// Dir は実行環境のOSに応じたキャッシュディレクトリを返す。
func Dir() (string, error) {
	return DirForOS(runtime.GOOS)
}

// DirForOS はgoos（runtime.GOOSと同じ値域）に応じたキャッシュディレクトリを返す。
//
// Goのruntime.GOOSはコンパイル時定数でありモックできない。そのため実装本体を
// goos引数で受け取る形に分離し、Dir()はruntime.GOOSを渡す薄いラッパーとする
// ことでテスト容易性を確保する（CLAUDE.mdの「外部依存は注入可能にする」
// 方針に沿った設計）。
func DirForOS(goos string) (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("ホームディレクトリを取得できません: %w", err)
	}

	switch goos {
	case "linux":
		if xdg := os.Getenv("XDG_CACHE_HOME"); xdg != "" {
			return filepath.Join(xdg, "mnemonic"), nil
		}

		return filepath.Join(home, ".cache", "mnemonic"), nil
	case "darwin":
		return filepath.Join(home, "Library", "Caches", "mnemonic"), nil
	case "windows":
		if localAppData := os.Getenv("LOCALAPPDATA"); localAppData != "" {
			return filepath.Join(localAppData, "mnemonic", "cache"), nil
		}

		return filepath.Join(home, "AppData", "Local", "mnemonic", "cache"), nil
	default:
		// Linux/Darwin/Windows以外のOSではXDG_CACHE_HOMEを見ずhome/.cacheに固定する。
		return filepath.Join(home, ".cache", "mnemonic"), nil
	}
}

// TemplateCachePath はversionに対応するテンプレートキャッシュパスを返す。
func TemplateCachePath(version string) (string, error) {
	dir, err := Dir()
	if err != nil {
		return "", err
	}

	return filepath.Join(dir, "templates", version), nil
}

// ClearCacheDir はcacheDir配下のキャッシュを削除する。
//
// templateOnly=falseの場合はKeystoreDirNameという名前のエントリを除くcacheDir直下の
// 全エントリを削除し、cacheDir自体は残す。templateOnly=trueの場合はtemplatesのみ
// 削除する。cacheDirが存在しない場合もエラーにはならない。
func ClearCacheDir(cacheDir string, templateOnly bool) error {
	if templateOnly {
		if err := os.RemoveAll(filepath.Join(cacheDir, "templates")); err != nil {
			return fmt.Errorf("キャッシュの削除に失敗しました: %w", err)
		}

		return nil
	}

	entries, err := os.ReadDir(cacheDir)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil
		}

		return fmt.Errorf("キャッシュの削除に失敗しました: %w", err)
	}

	for _, entry := range entries {
		// why not: 署名鍵まで消すと次回ビルドの署名鍵が変わり、既存APKへの上書き
		// インストールができなくなる（詳細はpipeline.createDebugKeystore）。
		if entry.Name() == KeystoreDirName {
			continue
		}

		if err := os.RemoveAll(filepath.Join(cacheDir, entry.Name())); err != nil {
			return fmt.Errorf("キャッシュの削除に失敗しました: %w", err)
		}
	}

	return nil
}

// InfoForDir はcacheDirのキャッシュ情報を取得する。
func InfoForDir(cacheDir string) (Info, error) {
	if _, err := os.Stat(cacheDir); err != nil {
		return Info{Directory: cacheDir}, nil
	}

	totalSize, err := dirSize(cacheDir)
	if err != nil {
		return Info{}, fmt.Errorf("キャッシュサイズの計算に失敗しました: %w", err)
	}

	version, expiresInDays := latestTemplateInfo(filepath.Join(cacheDir, "templates"))

	return Info{
		Directory:             cacheDir,
		SizeBytes:             totalSize,
		TemplateVersion:       version,
		TemplateExpiresInDays: expiresInDays,
	}, nil
}

func dirSize(root string) (int64, error) {
	var total int64

	err := filepath.WalkDir(root, func(_ string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		info, err := d.Info()
		if err != nil {
			return err
		}
		total += info.Size()

		return nil
	})
	if err != nil {
		return 0, err
	}

	return total, nil
}

// ReadTemplateMetadata はversionDir直下のメタデータを読む。無い・壊れている場合はok=false。
func ReadTemplateMetadata(versionDir string) (TemplateMetadata, bool) {
	data, err := os.ReadFile(filepath.Join(versionDir, TemplateMetadataFilename)) //nolint:gosec // キャッシュディレクトリ配下の固定ファイル名を読む用途のため妥当
	if err != nil {
		return TemplateMetadata{}, false
	}

	var metadata TemplateMetadata
	if err := json.Unmarshal(data, &metadata); err != nil {
		return TemplateMetadata{}, false
	}

	return metadata, true
}

// latestTemplateInfo はtemplateDir直下のバージョンのうちdownloaded_atが最も新しいものを
// 最新テンプレートとみなし、そのバージョン名とexpires_atまでの残り日数（切り上げ、
// 下限0）を返す。有効なメタデータを持つバージョンが無い場合は両方nilを返す。
//
// why not: ディレクトリの更新日時はダウンロード日時と一致せず、有効期間も
// --template-refresh-daysで変わるため、更新日時と固定日数からは求めない。
// ビルド自身が書いたメタデータを読むことで、ビルドのキャッシュ判定と一致させる。
// expires_atを解釈できない場合は、ビルドが無効なキャッシュとみなすのに合わせて0日とする。
func latestTemplateInfo(templateDir string) (*string, *int) {
	entries, err := os.ReadDir(templateDir)
	if err != nil {
		return nil, nil
	}

	var (
		latest       TemplateMetadata
		latestName   string
		latestLoaded time.Time
		found        bool
	)

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}

		metadata, ok := ReadTemplateMetadata(filepath.Join(templateDir, entry.Name()))
		if !ok {
			continue
		}

		downloadedAt, err := time.Parse(TemplateMetadataTimeLayout, metadata.DownloadedAt)
		if err != nil {
			continue
		}

		if !found || downloadedAt.After(latestLoaded) {
			latest = metadata
			latestName = entry.Name()
			latestLoaded = downloadedAt
			found = true
		}
	}

	if !found {
		return nil, nil
	}

	return new(latestName), new(daysUntil(latest.ExpiresAt))
}

func daysUntil(expiresAt string) int {
	expires, err := time.Parse(TemplateMetadataTimeLayout, expiresAt)
	if err != nil {
		return 0
	}

	return max(0, int(math.Ceil(time.Until(expires).Hours()/24)))
}
